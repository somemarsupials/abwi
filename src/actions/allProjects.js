import { api } from '../App';
import { ThunkGenerator } from '../lib';

const CONTEXT = 'PROJECTS';
const generator = new ThunkGenerator();

// actions

const actions = {
  TOGGLE_CREATE_MODAL: `${CONTEXT}/TOGGLE_CREATE_MODAL`
};

const SUPPORTED = ['FETCH', 'CREATE']

SUPPORTED.forEach(function(method) {
  Object.assign(actions, generator.actions(CONTEXT, method));
});

export { actions };

// requests

async function fetchRequest() {
  return await fetch(`${api}/projects`);
};

async function createRequest(params) {
  let config = { 
    method: 'post', 
    body: JSON.stringify(params) 
  };
  return await fetch(`${api}/projects`, config);
};

// action generators

export function toggleCreateProjectModal() {
  return {
    type: actions.TOGGLE_CREATE_MODAL,
  };
};

// thunks

export const fetchProjects = 
  generator.generate(fetchRequest, CONTEXT, 'FETCH');

export const createProjects = 
  generator.generate(createRequest, CONTEXT, 'CREATE');
