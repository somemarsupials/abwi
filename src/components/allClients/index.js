export { default as ClientsPage } from './clientsList';
export { default as ClientsList } from './clientsPage';
