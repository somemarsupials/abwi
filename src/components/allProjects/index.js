export { default as List } from './list';
export { default as Page } from './page';
