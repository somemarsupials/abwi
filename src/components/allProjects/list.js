import React from 'react';
import { Link } from 'react-router-dom';

export default function(props) {
  let projects = props.projects || [];

  return(
    <table className="table">
      <thead>
        <tr>
          <th>Name</th>
        </tr>
      </thead>
      <tbody>
        { projects.map(function(project, index) {
          return (
            <tr key={index}>
              <td>
                <Link to={`/projects/${project.id}`}>{project.title}</Link>
              </td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};
