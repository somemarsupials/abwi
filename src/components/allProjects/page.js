import React from 'react';

import { TitleBar, ErrorBar, InfoBar } from '../common/';
import List from './list';

export default function(props) {
  let error = props.error || !props.projects;

  return (
    <div>
      <TitleBar text={'Projects'} /> 
      { error &&
        <ErrorBar text={'Could not find projects'} detail={props.error} />
      }
      { !error && props.fetching && 
        <InfoBar text={'Fetching projects...'} />
      }
      { !error && !props.fetching &&
        <List projects={props.projects} />
      }
    </div>
  );
};
