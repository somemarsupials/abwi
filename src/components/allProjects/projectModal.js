import React from 'react';
import { Modal, Button } from 'react-bootstrap';

export default function(props) {
  return (
    <Modal onHide={props.onHide} show={props.show}>
      <Modal.Header>
        <h4>Create project</h4>
      </Modal.Header>
      <Modal.Body>
        
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={props.onHide}>Close without saving</Button>
      </Modal.Footer>
    </Modal>
  );
};
