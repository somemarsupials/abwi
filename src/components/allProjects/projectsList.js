import React from 'react';
import { Link } from 'react-router-dom';

export default function(props) {
  if (!props.projects || props.projects.length === 0) {
    return (<span>No projects found</span>);
  };

  return(
    <table className="table">
      <thead>
        <tr>
          <th>Name</th>
        </tr>
      </thead>
      <tbody>
        { props.projects.map(function(project, index) {
          return (
            <tr key={index}>
              <td>
                <Link to={`/projects/${project.id}`}>{project.title}</Link>
              </td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};
