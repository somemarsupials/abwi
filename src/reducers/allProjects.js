import { combineReducers } from 'redux';
import { actions } from '../actions/allProjects';

function fetching(state = false, action) {
  switch (action.type) {
    case actions.FETCH_LOADING:
      return action.isLoading;
    default:
      return state;
  }
};

function fetchSuccess(state = [], action) {
  switch (action.type) {
    case actions.FETCH_SUCCESS:
      return action.result;
    default:
      return state;
  }
};

function fetchFail(state = null, action) {
  switch (action.type) {
    case actions.FETCH_FAIL:
      return action.error;
    default:
      return state;
  }
};

function createModal(state = false, action) {
  switch (action.type) {
    case actions.TOGGLE_CREATE_MODAL:
      return action.active;
    default:
      return state;
  }
};

export default combineReducers({
  fetching: fetching,
  data: fetchSuccess,
  error: fetchFail,
  createModal: createModal,
});
