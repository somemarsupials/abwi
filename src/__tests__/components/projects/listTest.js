import React from 'react';
import { shallow } from 'enzyme';

import { List } from '../../../components/allProjects';

describe('List', () => {
  let list;
  let items;

  beforeEach(() => {
    items = ['item 1', 'item 2']
  });

  describe('without items', () => {
    beforeEach(() => {
      list = shallow(<List />);
    });

    it('renders with error bar', () => {
      expect(list).toMatchSnapshot();
    });
  });

  describe('without items', () => {
    beforeEach(() => {
      list = shallow(<List items={items} />);
    });

    it('renders with items', () => {
      expect(list).toMatchSnapshot();
    });
  });
});
