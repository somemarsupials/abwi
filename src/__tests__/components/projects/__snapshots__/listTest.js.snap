// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List without items renders with error bar 1`] = `
<span>
  No projects found
</span>
`;

exports[`List without items renders with items 1`] = `
<span>
  No projects found
</span>
`;
